### ISSUE TITLE
Scheduler should consider the current status of cpu/memory of hosts while scheduling.

### ISSUE THREAD

**Issue by hardikdr** [kubernetes/kubernetes#12455]
Hi all..
I have recently started looking into kubernetes project and strongly willing to contribute to the project. I have environment set up of Kubernetes on CoreOS baremetal servers. 
It seems to me that scheduler is not considering the current state of resources(cpu/memory) of hosts while taking decisions.

I performed following steps:
1. There are 2 hosts(slaves), host1 and host2. And one host0(master) which consists of scheduler and apiserver.
2. I started creating pods , in the kubernetes cluster by just changing the "name" field in pod template everytime. 
[I did not mention cpu and memory limits in the pod template.
Pod had just one simple container (ubuntu image).]
The scheduler was nicely putting pods on different hosts one by one.
3. Then I intentionally exhausted the cpu of host1 by starting some processes on host1 via pod.
4. Now on adding new pods, the scheduler was still scheduling the pods on the host1 as if it is not considering its current cpu status, while it should put new pod on host2 which is a better decision.

I understand it may possible that in near future the cpu usage may go down in host1 . There may be other issues as well. 
But , I think scheduler should consider the current status of cpu/memory while scheduling the pod. I am not sure but scheduler can utilize the data from cadvisor.
Please let me know if I am missing something.

**Comment by a-robinson** [comment]
Yes, this is very good feedback! There's a proposal under review right now in #11951 for how to take this into account. I'm going to close this in favor of that proposal, but feel free to reopen if you think there are separate things to be discussed here.

**Comment by davidopp** [comment]
This isn't exactly the same issue; #11951 is about exposing the metrics to the API server, while I think @HardikDR is talking about using those metrics when scheduling.

Our initial thought was to use it for scheduling best-effort pods only (schedule to machine with lowest recent load). But it won't be possible to implement this until the node metrics are surfaced to the API server, so that the scheduler can read them.

@mikedanese

**Comment by dalanlan** [comment]
The related doc says as following,
>Second, it applies a set of "priority functions" that rank the nodes that weren't filtered out by the predicate check. For example, it tries to spread Pods across nodes while at the same time favoring the least-loaded nodes (where "load" here is sum of the resource limits of the containers running on the node, divided by the node's capacity).
Let's say the so-called **load**  does not take cpu/mem into account, what exactly does it refer to (if i may ask)?

**Comment by vishh** [comment]
#11951

**Comment by HardikDR** [comment]
@dalanlan 
Load does take cpu/mem into account . 
But that cpu/mem is something that we provide as resource limits in pod template. The resource limits specifies the max resource that might be requested by the container.

**Comment by vishh** [comment]
cc @mvdan who is planning to work on this, once the metrics API is completed.

**Comment by dalanlan** [comment]
Hmm.. Let me try to make myself clear.
What concerns me much is the second step of ranking nodes with "priority function", not the first step(which as you pointed out, does the jobs to filter out improper nodes).
So again, what does load refer to when ranking nodes?

**Comment by davidopp** [comment]
> what does load refer to when ranking nodes?

Sum of the Limits as specified here
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/types.go#L717

This is in the process o changing to sum of Requests, but same idea.

In other words: load is a currently a *static* concept defined by what people specified when they created their Pods/containers.

In the future it will also have a *dynamic* component based on looking at actual resource utilization on the machine.

**Comment by dalanlan** [comment]
Many thanks. Looking fwd to then.

**Comment by HardikDR** [comment]
I am really interested in working on this !! 
Please guide.
@vishh

**Comment by HardikDR** [comment]
I am interested in working on this !!
Please guide.
@vishh

**Comment by vishh** [comment]
@HardikDR: #11951 has to be completed first. I can update this thread once that is done.

**Comment by HardikDR** [comment]
@vishh  Okay.. Thanks.

**Comment by dalanlan** [comment]
@HardikDR  
#11951 has been merged earlier. FYI.

**Comment by dalanlan** [comment]
Is there anyone working around this, if i may ask?

**Comment by davidopp** [comment]
It's blocked on #12952

**Comment by HardikDR** [comment]
#12952   has to be merged first.

**Comment by bgrant0607** [comment]
This is a dupe of #274

**Comment by davidopp** [comment]
Good point. But almost all the discussion on that issue is from a year and a half ago, so I'm going to close that one in favor of this one.

**Comment by HardikDR** [comment]
Is heapster-based solution reliable for this purpose  ?

**Comment by HardikDR** [comment]
@davidopp  @vishh #12952 is closed now. 
1. Are we waiting on some other issue ? 
2. Why can't we just use the recent metrics from Heapster http endpoint.
    Eg: heapster/api/v1/model/nodes
3. Can we add 2 more metrics , DiskIO and NetworkIO , taking it from cAdvisor of all nodes(per-node stats is not available in Heapster). It can be used for better scheduling.
Please guide.

**Comment by davidopp** [comment]
#12952 is closed but wasn't implemented. As the last comment there says, "Closing in favor of the new metrics API effort." (https://github.com/kubernetes/kubernetes/pull/12952#issuecomment-172804245) The new metrics API effort is described in #18824 (sorry that issue didn't explain -- I added a comment just now pointing to #18824).

We were hoping to get the metrics API into 1.2 but we ran out of time. It will definitely be in 1.3. We also hope to do usage-based scheduling for best-effort pods in 1.3, since the metrics API will hopefully be done early in the 1.3 release cycle. 

I'll let @vishh and @piosz comment on (3) as I am not sure exactly wich metrics will be exposed.

**Comment by piosz** [comment]
No sure what do you mean by
>per-node stats is not available in Heapster

All rate metrics are aggregated on node level. Also all metrics that are exposed by cadvisor on node level are being scrapped by Heapster. Some kind of disk and network metrics are exported by Heapster, please take a look into [metrics.go](https://github.com/kubernetes/heapster/blob/heapster-scalability/metrics/core/metrics.go#L197) to verify if this is what you want. Disk and network metric will be available in Metrics API once disk and network will become first class resource in Kubernetes (not sure what is the plan and the current state).

cc @mwielgus

**Comment by AudriusButkevicius** [comment]
Are there any blockers left for this?
Also, what's the relationship between this and #18438?
I guess both of them are talking about the same problem?

**Comment by anfernee** [comment]
I would like to take care of it. 

(not sure if this would work...)
/assign @anfernee

**Comment by bgrant0607** [comment]
See #18620, and please discuss in SIG scheduling first.
